api/filters/reference_instance_filter.go

```diff
@@ -0,0 +1,217 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package filters
+
+import (
+	"context"
+	"fmt"
+	"github.com/Peripli/service-manager/constant"
+	"github.com/Peripli/service-manager/pkg/log"
+	"github.com/Peripli/service-manager/pkg/query"
+	"github.com/Peripli/service-manager/pkg/types"
+	"github.com/Peripli/service-manager/pkg/util"
+	"github.com/Peripli/service-manager/storage"
+	"github.com/tidwall/gjson"
+	"github.com/tidwall/sjson"
+	"net/http"
+
+	"github.com/Peripli/service-manager/pkg/web"
+)
+
+const (
+	ReferenceInstanceFilterName        = "ReferenceInstanceFilter"
+	ReferenceParametersAreNotSupported = "Reference service instance parameters are not supported"
+	contextKey                         = "context"
+)
+
+// serviceInstanceVisibilityFilter ensures that the tenant making the provisioning/update request
+// has the necessary visibilities - i.e. that he has the right to consume the requested plan.
+type referenceInstanceFilter struct {
+	repository       storage.TransactionalRepository
+	tenantIdentifier string
+}
+
+func NewReferenceInstanceFilter(repository storage.TransactionalRepository, tenantIdentifier string) *referenceInstanceFilter {
+	return &referenceInstanceFilter{
+		repository:       repository,
+		tenantIdentifier: tenantIdentifier,
+	}
+}
+
+func (*referenceInstanceFilter) Name() string {
+	return ReferenceInstanceFilterName
+}
+
+func (f *referenceInstanceFilter) Run(req *web.Request, next web.Handler) (*web.Response, error) {
+	switch req.Request.Method {
+	case http.MethodPost:
+		return f.handleProvision(req, next)
+	case http.MethodPatch:
+		return f.handleServiceUpdate(req, next)
+	}
+	return nil, nil
+}
+
+func (*referenceInstanceFilter) FilterMatchers() []web.FilterMatcher {
+	return []web.FilterMatcher{
+		{
+			Matchers: []web.Matcher{
+				web.Path(web.ServiceInstancesURL + "/**"),
+				web.Methods(http.MethodPost, http.MethodPatch),
+			},
+		},
+	}
+}
+
+func (f *referenceInstanceFilter) handleProvision(req *web.Request, next web.Handler) (*web.Response, error) {
+	ctx := req.Context()
+	servicePlanID := gjson.GetBytes(req.Body, planIDProperty).String()
+	isReferencePlan, err := f.isReferencePlan(ctx, servicePlanID)
+	if err != nil {
+		return nil, err
+	}
+	if !isReferencePlan {
+		return next.Handle(req)
+	}
+
+	// Ownership validation
+	path := fmt.Sprintf("%s.%s", contextKey, f.tenantIdentifier)
+	callerTenantID := gjson.GetBytes(req.Body, path).String()
```

Comment:
This does not really check ownership? anyone can alter the request.body details?
to validate the tenantId, something like this can be used:
tenantID := query.RetrieveFromCriteria(labelKey, query.CriteriaForContext(req.Context())...)

instance ownership should be checked as part of the instance ownership plugin/ filter.
do we really need this extra validation here?